src/main/java/io/javalin/core/compression/DynamicCompressionHandler.kt

```diff
@@ -0,0 +1,60 @@
+package io.javalin.core.compression
+
+import io.javalin.core.JavalinConfig
+import io.javalin.core.util.Header
+import io.javalin.http.Context
+import javax.servlet.http.HttpServletResponse
+
+/**
+ * Class designed to handle dynamic content compression for Javalin.
+ *
+ * Uses DynamicCompressionStrategy within JavalinConfig to determine the type and level of compression to be applied.
+ *
+ * @see DynamicCompressionStrategy
+ */
+class DynamicCompressionHandler(val ctx: Context, val config: JavalinConfig) {
+
+    private val compressionStrategy = config.inner.dynamicCompressionStrategy
+
+    /**
+     * @param res The response that we wish to encode
+     */
+    fun compressResponse(res: HttpServletResponse) {
+        val resultStream = ctx.resultStream()!!
+        when {
+            brotliShouldBeDone() -> { //Do Brotli
+                res.setHeader(Header.CONTENT_ENCODING, "br")
+                compressionStrategy.brotli?.write(res.outputStream, resultStream.readBytes())
+                return
+            }
+            gzipShouldBeDone() -> { //Do GZIP
+                res.setHeader(Header.CONTENT_ENCODING, "gzip")
+                compressionStrategy.gzip?.write(res.outputStream, resultStream.readBytes())
+                return
+            }
+            else -> {
+                resultStream.copyTo(res.outputStream) //No compression
+            }
+        }
+    }
+
+    private fun resultExceedsMtu(): Boolean {
+        return ctx.resultStream()?.available() ?: 0 > 1500 // mtu is apparently ~1500 bytes
+    }
+
+    private fun supportsEncoding(encoding: String): Boolean {
+        return (ctx.header(Header.ACCEPT_ENCODING) ?: "").contains(encoding, ignoreCase = true)
+    }
+
+    private fun gzipShouldBeDone(): Boolean {
+        return  compressionStrategy.gzip != null
+                && resultExceedsMtu()
+                && supportsEncoding("gzip")
+    }
+
+    private fun brotliShouldBeDone(): Boolean {
+        return  compressionStrategy.brotli != null
+                && resultExceedsMtu()
+                && supportsEncoding("br")
+    }
```

Comment:
You could these four private methods to function expressions:

```kotlin
private fun resultExceedsMtu() = ctx.resultStream()?.available() ?: 0 > 1500 // mtu is apparently ~1500 bytes
```

etc.

Reply:
I was wondering about that, actually, since I am new to Kotlin. What exactly are the benefits of this "function expression" notation? 

The reasons I tend to avoid it is because 1) they make it harder to work with debugger breakpoints, and 2) they make it hard to see what the return type of the function is, which again makes debugging more difficult.

But again, I am new to Kotlin and very much used to the strongly typed variables. I may be thinking along the entirely wrong set of tracks here, which is why I am asking. :)

Reply:
> What exactly are the benefits of this "function expression" notation?

For me it's readability. If it's a function expression I know that it's a small piece of code that only does one thing.

> they make it harder to work with debugger breakpoints, 

Can't argue with that one. I've found myself rewriting sometimes to use standard syntax for debugging, but I still prefer expression syntax for reading code.

> they make it hard to see what the return type of the function is, which again makes debugging more difficult.

You could still keep the explicit type declaration with function expressions:

```kotlin
private fun resultExceedsMtu(): Boolean = ctx.resultStream()?.available() ?: 0 > 1500 // mtu is apparently ~1500 bytes
```

Kotlin is as strongly typed as Java, it just uses inference more often. 

> I may be thinking along the entirely wrong set of tracks here, which is why I am asking. :)

No, your thinking is fine. I don't mind leaving them as normal functions, just wanted to point out that you could write them as expressions (it's something a lot of people who are new to Kotlin find a bit weird).

Reply:
> You could still keep the explicit type declaration with function expressions

Didn't think of that! Brilliant, best of both worlds :D 

> just wanted to point out that you could write them as expressions

I appreciate that. Always good to know, especially for a beginner!